feat: add refunds GraphQL query

kieckhafer opened the pull request:
Part of #5158 
Impact: **minor**  
Type: **feature**

A new `listRefunds` query allows you to query refunds of payments on a particular order. This was previously a meteor only method. The meteor method will be removed with #5158.

## Breaking changes
None

## Testing
1. Place an order using both the Stripe payment method and IOU method
1. Create a refund for each type of payment
1. Use the new `listRefunds` query to see the refunds

```
listRefunds(orderId: {ORDER-ID}, shopId: {SHOP-ID}) {
   _id
    amount
    ...otherFieldsYouWant
  }
```

I've added tests here to test the permission / requirements of variables, but did not add tests to check a result, as the payment method has it's own tests for this element: https://github.com/reactioncommerce/reaction/blob/feat-kieckhafer-listRefundsGQLQuery/imports/plugins/included/payments-stripe/server/no-meteor/util/stripeListRefunds.test.js#L10

kieckhafer commented:
@aldeed I wasn't sure if this should go into `Payments` or `Orders`. I put it in `Orders` because each specific payment method has it's own way of calculating the refunds, and this is just a list that is displayed on the `Orders` operator UI.

What do you think about it's location?

aldeed commented:
@kieckhafer I think I have the same thoughts as you, could go either place. Since payments are currently coupled with orders, leaving it in orders seems fine.

I'm assigning @focusaurus to do the full code review here, but my only comment is that for GraphQL, we try to omit words like `get` and `list` from the query names. That's pretty standard GQL best practice. So I'd go with just `refunds`.

Also keep in mind that something like this fits pretty well into the payment resolver. In other words, rather than (or in addition to) `Query.refunds` resolver, you could have a `Payment.refunds` resolver. This is nice because you'd already know order ID and payment ID in that resolver, plus you could avoid doing an extra query in Catalyst.

```graphql
{
  order {
    payments {
      refunds {
        amount
      }
    }
  }
}
```

Or for a combined list for all payments, it could be similarly done as `Order.refunds` resolver. There can be several resolvers that all call the same query function giving clients various ways to get the same info.

kieckhafer commented:
@aldeed makes sense. I took the pattern of `listRefunds` from the plugins that do it, we currently have `stripleListRefunds`. Is it OK if it's not the first word?

focusaurus commented on imports/plugins/core/orders/server/no-meteor/queries/refunds.js:
```diff
@@ -0,0 +1,50 @@
+
+/**
+ * @name refunds
+ * @method
```
How/why are we using `@method` here? I see we're using it all over the place, it's just not clear from the docs what it's supposed to convey or how it renders in the doc site.

kieckhafer replied:
This is a habit carried over from using `meteor` methods. For certain methods, they wouldn't appear in jsdoc without this. We can probably stop using it, unless anyone has any objections / reasons why we do still use it? @machikoyasuda @mikemurray any ideas?

machikoyasuda replied:
As long as the method's docs appear in http://api.docs.reactioncommerce.com/, we're good 👍 

Back in the old days, some Meteor methods would not appear w/o the `@method`. You can test it out locally by running `npm run docs` and checking the HTML files that get created in your personal `tmp/reaction-docs` folder to see if that method was rendered.

focusaurus commented on imports/plugins/core/orders/server/no-meteor/queries/refunds.js:
```diff
@@ -0,0 +1,50 @@
+  if (Array.isArray(order.payments)) {
+    for (const payment of order.payments) {
+      /* eslint-disable no-await-in-loop */
+      const shopRefunds = await getPaymentMethodConfigByName(payment.name).functions.listRefunds(context, payment);
```
Can we make this loop run in parallel by creating the promises without awaiting them then `await Promise.all` after the loop?

kieckhafer replied:
I think the way I'm doing it has been the pattern we've been using with the non-meteor code. In Meteor code we were using the `Promise.await`, but in all our non-meteor code we are doing the await inside the loop. @aldeed is my assessment correct? is there a reasoning behind why we do it with `await` in the loop more recently?

aldeed replied:
@kieckhafer Actually the pattern is to `await Promise.all` whenever parallel is fine. In many places, parallel is not fine because the order of execution matters or it's transformation chain, and in those instances we `await` in the for loop.

As I typed this it started to sound familiar, and then I remembered this: https://docs.reactioncommerce.com/docs/dev-how-do-i#loop-over-async-function-results

aldeed replied:
In this specific case, it seems to me that parallel would be good.

kieckhafer replied:
A little confused about how to go about this.

I need to use the results from line 38 on line 40, where would I `await Promise.all`? If I change line 36 to a map, I still need the result from the await inside the map, I can't wait until after it to resolve the promises.

focusaurus replied:
Here's what I think I'd do. It's not going to win any beauty pageants but I think it works in parallel and gets you the combination you need of `shopRefunds` and the corresponding `payment` in the same scope. This is from `const paymentRefunds = [];` to the end of the function body.

```js
  const paymentRefunds = [];

  if (Array.isArray(order.payments)) {
    const shopRefundsWithPaymentPromises = order.payments.map((payment) =>
      getPaymentMethodConfigByName(payment.name)
        .functions.listRefunds(context, payment)
        .then((shopRefunds) => shopRefunds, payment));
    // This should give us an arary of objects with properties
    // `shopRefunds` and the corresponding `payment`
    const shopRefundsWithPayments = await Promise.all(shopRefundsWithPaymentPromises);
    shopRefundsWithPayments.forEach(({ shopRefunds, payment }) => {
      const shopRefundsWithPaymentId = shopRefunds.map((shopRefund) => ({
        ...shopRefund,
        paymentId: payment._id,
        paymentDisplayName: payment.displayName
      }));
      paymentRefunds.push(...shopRefundsWithPaymentId);
    });
  }

  return paymentRefunds;

```

kieckhafer replied:
Got it, this makes sense. I wasn't wrapping my head around that second loop in the `forEach`.

focusaurus commented on imports/plugins/core/orders/server/no-meteor/queries/refunds.test.js:
```diff
@@ -0,0 +1,51 @@
+  })).rejects.toThrowErrorMatchingSnapshot();
+
+  expect(mockContext.userHasPermission).toHaveBeenCalledWith(["orders", "order/fulfillment", "order/view"], "SHOP_ID");
+});
```
Can we add a base success test case?

kieckhafer replied:
added

focusaurus commented on imports/plugins/core/orders/server/no-meteor/resolvers/Refund/createdAt.js:
```diff
@@ -0,0 +1,17 @@
+  const formattedDate = moment(refund.created).format();
+
+  return formattedDate;
+}
```
Can we add a quick unit test for this? Almost trivial but easy to test. Any concerns about stripe specifics in a generic plugin?

kieckhafer replied:
We could do it in the stripe plugin instead, however the `moment.format()` will accept other date formats as well and spit them out the same way, so although my comment and the reasoning behind this is because of stripe, this is really just making a consistent format for all dates that we get back.

focusaurus commented on imports/plugins/included/marketplace/server/no-meteor/util/stripeListRefunds.js:
```diff
@@ -25,11 +25,13 @@ export default async function stripeListRefunds(context, paymentMethod) {
       result.push({
-        type: refund.object,
+        _id: refund.id,
         amount: refund.amount / 100,
```
I know this line didn't change but are we USD-only in stripe at this point? Is it safe to do this conversion regardless of `refund.currency`?

kieckhafer replied:
Great question. It's probably not safe to do in all currencies: https://stripe.com/docs/currencies#zero-decimal

@spencern are we USD-only for Stripe?

spencern replied:
We should not assume USD only.

@kieckhafer will you create a ticket to audit the stripe integration to see where we have incompatibility with other currencies?

kieckhafer replied:
Issue created: https://github.com/reactioncommerce/reaction/issues/5383

kieckhafer commented:
@focusaurus I added a new query here, `refundsByPaymentId`, and also added a couple resolvers so that you can query `refunds` on an order, and also `refunds` per payment, so there is some additional functionality to check since your last review.

```
query {
  orderByReferenceId(id: "ORDER-ID", shopId: "SHOP-IS") {
    _id
    payments {
      _id
      method {
        name
      }
      refunds {
        amount {
          displayAmount
        }
      }
    }
    refunds {
       amount {
        displayAmount
      }
      paymentDisplayName
    }
    shop {
      _id
    }
  }
}
```

kieckhafer commented on imports/plugins/core/orders/server/no-meteor/resolvers/Order/payments.js:
```diff
@@ -0,0 +1,34 @@
+    return order.payments.map(async (payment) => {
+      const xformPayment = xformOrderPayment(payment);
+
+      const refunds = await context.queries.refundsByPaymentId(context, {
```
The reason I did this here, instead of inside the `xformOrderPayment` function, is I wanted to keep all the refunds xforms and data inside this orders package, instead of in the `core` package where `xformOrderPayment` is.

This could be moved into into  `xformOrderPayment` if that seems like a better place.

focusaurus commented on imports/plugins/core/orders/server/no-meteor/resolvers/Query/refundsByPaymentId.js:
```diff
@@ -0,0 +1,29 @@
+ * @param {Object} _ - unused
+ * @param {ConnectionArgs} args - An object of all arguments that were sent by the client
+ * @param {String} args.orderId - ID of the order
+ * @param {String} args.orderId - ID of the payment
```
Copypasta: `args.paymentId`

kieckhafer replied:
🍝

aldeed commented on imports/plugins/core/orders/server/no-meteor/resolvers/Refund/createdAt.js:
```diff
@@ -0,0 +1,17 @@
+export default function createdAt(refund) {
+  // graphql-iso-data won't accept the Unix timestamp that is provided by Stripe
+  // we need to use moment to convert it to a readable format
+  const formattedDate = moment(refund.created).format();
```
I believe `created` is MS so you can do `const formattedDate = new Date(refund.created);` without needing Moment.

aldeed commented on imports/plugins/core/orders/server/no-meteor/resolvers/Refund/index.js:
```diff
@@ -0,0 +1,10 @@
+  amount,
+  createdAt,
+  paymentDisplayName: (node) => node.paymentDisplayName,
+  reason: (node) => node.reason
```
Neither of these last two props should be necessary. If GraphQL doesn't find a resolver for it, it will just use the object property with that name, exactly what you're doing here.

kieckhafer replied:
Does that mean `_id` also isn't necessary?

kieckhafer replied:
nevermind, commented before I saw your other comment below.

aldeed commented on imports/plugins/core/orders/server/no-meteor/resolvers/Refund/index.js:
```diff
@@ -0,0 +1,10 @@
+import createdAt from "./createdAt";
+
+export default {
+  _id: (node) => node._id,
```
You need to encode this ID here. May need a new transform for "reaction/refund" namespace.

You should also add `paymentId` here and encode that (and add it to GraphQL schema).

aldeed commented on imports/plugins/core/orders/server/no-meteor/schemas/schema.graphql:
```diff
@@ -395,6 +410,27 @@ type OrderNote {
+  reason: String
+
+  "The type of refund"
+  type: RefundType
```
Do we use/show this anywhere? It's unclear to me what it's for so maybe we should omit it from the GraphQL for now?

kieckhafer replied:
No, we don't use it, will remove.

kieckhafer commented:
@aldeed all comments addressed, ready for another check

kieckhafer commented:
@aldeed comments addressed, ready for a final look
